angel_system/data/common/load_data.py

```diff
@@ -174,14 +187,292 @@ def activities_from_dive_csv(filepath: str) -> List[Activity]:
+                next_activity = gt[a_id + 1]
+                next_sub_step_str = next_activity.class_label
+
+                if "(step " not in next_sub_step_str:
+                    continue
+                next_step = next_sub_step_str.split("(")[1][:-1]
+
+                inter_class = sanitize_str(f"In between {step} and {next_step}")
+
+                gt.append(
+                    Step(
+                        labels.index(inter_class),
+                        inter_class,
+                        activity.end,
+                        next_activity.start,
+                        1,
+                        True,
+                    )
+                )
+
+    if add_before_after_steps:
+        ns = "not started"
+        gt.append(
+            Step(
+                labels.index(ns),
+                ns,
+                min_start_time,
+                first_activity.start,
+                1,
+                True,
+            )
+        )
+
+        f = "finished"
+        gt.append(
+            Step(
+                labels.index(f),
+                f,
+                last_activity.end,
+                max_end_time,
+                1,
+                True,
+            )
+        )
+
+    return gt
+
+
+def steps_from_ros_export_json(filepath: str) -> Tuple[List[str], List[Activity]]:
+    """
+    Load a number of predicted steps from a JSON file that is the result
+    of bag extraction of a `TaskUpdate.msg` message topic.
+
+    Class labels are converted to lower-case and stripped of any extraneous
+    whitespace.
+
+    See message file located here: ros/angel_msgs/msg/TaskUpdate.msg
+    See bag extraction conversion logic located here: ros/angel_utils/scripts/bag_extractor.py
+
+    :param filepath: Filesystem path to the JSON file.
+    :return: List activity predicted labels and a list of loaded activity
+        annotations (predicted).
+    """
+    log.info(f"Loading predicted steps from: {filepath}")
+    with open(filepath, "r") as f:
+        data = json.load(f)
+    step_seq: List[Step] = []
+    for step_json in data:
+        # This activity window start/end times in seconds.
+        time_stamp = float(step_json["header"]["time_sec"]) + (
+            float(step_json["header"]["time_nanosec"]) * 1e-9
+        )
+        latest_sensor_ts = float(step_json["latest_sensor_input_time"])
+        current_step_id = step_json["current_step_id"]
+        current_step = step_json["current_step"]
+        conf_vec = step_json["hmm_step_confidence"]
+        completed_vec = step_json["completed_steps"]
+
+        step_seq.append(
+            Step(
+                current_step_id,  # current id
+                sanitize_str(current_step),  # class_label
+                latest_sensor_ts,  # start
+                latest_sensor_ts,  # end
+                conf_vec[current_step_id],  # conf
+                completed_vec[current_step_id],  # completed
+            )
+        )
```

Comment:
Not every message is a unique step, the output here is a streaming state message. I believe you will need to detect transitions if you want to have your output be discrete step time volumes.

Reply:
OBE